openmdao/core/driver.py

```diff
@@ -22,7 +23,140 @@
+        The number of times the total jacobian was computed.
+    deriv_time : float
+        The time spent computing the total jacobian.
+    exit_status : str
+        A string that may provide more detail about the results of the driver run.
+    success : bool
+        A boolean that dictates whether or not the driver was successful.
+    """
+
+    def __init__(self):
+        """
+        Initialize the DriverResult object.
+        """
+        self.runtime = 0.0
+        self.iter_count = 0
+        self.obj_calls = 0
+        self.obj_time = 0.0
+        self.deriv_calls = 0
+        self.deriv_time = 0.0
+        self.exit_status = 'NOT_RUN'
+        self.success = False
+
+    def reset(self):
+        """
+        Set the driver result attributes back to their default value.
+        """
+        self.runtime = 0.0
+        self.iter_count = 0
+        self.obj_calls = 0
+        self.obj_time = 0.0
+        self.deriv_calls = 0
+        self.deriv_time = 0.0
+        self.exit_status = 'NOT_RUN'
+        self.success = False
+
+    def __getitem__(self, s):
+        """
+        Provide key access to the attributes of DriverResult.
+
+        This is included for backward compatibility with some
+        tests which require dictionary-like access.
+
+        Parameters
+        ----------
+        s : str
+            The name of the attribute.
+
+        Returns
+        -------
+        object
+            The value of the attribute
+        """
+        return getattr(self, s)
+
+    def __bool__(self):
+        """
+        Mimick the behavior of the previous `failed` return value of run_driver.
+
+        The return value is True if the driver was NOT successful.
+        An OMDeprecationWarning is currently issued so users know to change their code.
+        Users should utilize the `success` attribute to test for driver success.
+
+        Returns
+        -------
+        bool
+            True if the Driver was NOT successful.
+
+        """
+        issue_warning(msg='boolean evaluation of DriverResult is temporarily implemented '
+                      'to mimick the previous `failed` return behavior of run_driver.\n'
+                      'Use the `success` attribute of the returned DriverResult '
+                      'object to test for successful driver completion.',
+                      category=OMDeprecationWarning)
+        return not self.success
+
+    @staticmethod
+    def track_stats(kind):
+        """
+        Decorate methods to tracks either the objective the deriv time and call count for a Driver.
+
+        This decorator should be applied to the _objfunc or _gradfunc (or equivalent) methods
+        of drivers. It will either accumulated the elapsed time in driver.result.obj_time or
+        driver.result.deriv_time, based on the value of time_type.
+
+        Parameters
+        ----------
+        kind : str
+            One of 'obj' or 'deriv', specifying which statistics should be accumulated.
```

Comment:
maybe pedantic, but i don't know about this being called 'obj', since you don't always have an objective (and, when you do want an objective, you want the constraints too, which also come along for free.) I know that I can't just say this without offerig some alternatives ... 

"compute" , "compute_totals"
"function", "gradient"  (note, this allows for our future "hessian")
"eval", "deriv_eval"